src/custom/components/Header/MenuTree/index.tsx

```diff
@@ -0,0 +1,165 @@
+import {
+  MAIN_MENU,
+  MenuTreeItem,
+  MenuItemKind,
+  InternalLink,
+  ExternalLink,
+  DropDownItem,
+  MenuLink,
+} from 'constants/mainMenu'
+import { ExternalLink as ExternalLinkComponent } from 'theme/components'
+
+// Assets
+import IMAGE_MOON from 'assets/cow-swap/moon.svg'
+import IMAGE_SUN from 'assets/cow-swap/sun.svg'
+
+interface MenuImageProps {
+  title: string
+  iconSVG?: string
+  icon?: string
+}
+
+function MenuImage(props: MenuImageProps) {
+  const { title, iconSVG, icon } = props
+
+  if (iconSVG) {
+    return <SVG src={iconSVG} description={`${title} icon`} />
+  } else if (icon) {
+    return <img src={icon} alt={`${title} icon`} />
+  } else {
+    return null
+  }
+}
+
+interface InternalExternalLinkProps {
+  link: InternalLink | ExternalLink
+  handleMobileMenuOnClick: () => void
+}
+
+function InternalExternalLink({ link, handleMobileMenuOnClick }: InternalExternalLinkProps) {
+  const { kind, title, url, iconSVG, icon } = link
+  const menuImage = <MenuImage title={title} icon={icon} iconSVG={iconSVG} />
+  const isExternal = kind === MenuItemKind.EXTERNAL_LINK
+
+  if (isExternal) {
+    return (
+      <ExternalLinkComponent href={url} onClickOptional={handleMobileMenuOnClick}>
+        {menuImage}
+        {title}
+      </ExternalLinkComponent>
+    )
+  } else {
+    return (
+      <StyledNavLink exact to={url} onClick={handleMobileMenuOnClick}>
+        {menuImage}
+        {title}
+      </StyledNavLink>
+    )
+  }
+}
+
+interface ContextProps {
+  darkMode: boolean
+  toggleDarkMode: () => void
+  handleMobileMenuOnClick: () => void
+}
+
+type DarkModeButtonProps = {
+  context: ContextProps
+}
+
+function DarkModeButton({ context }: DarkModeButtonProps) {
+  const { darkMode, toggleDarkMode, handleMobileMenuOnClick } = context
+  const description = `${darkMode ? 'Sun/light' : 'Moon/dark'} mode icon`
+  const label = (darkMode ? 'Light' : 'Dark') + ' Mode'
+  return (
+    <button
+      onClick={() => {
+        handleMobileMenuOnClick()
+        toggleDarkMode()
+      }}
+    >
+      <SVG src={darkMode ? IMAGE_SUN : IMAGE_MOON} description={description} /> {label}
+    </button>
+  )
+}
+
+interface LinkProps {
+  link: MenuLink
+  context: ContextProps
+}
+
+function Link({ link, context }: LinkProps) {
+  switch (link.kind) {
+    case MenuItemKind.DARK_MODE_BUTTON:
+      return <DarkModeButton context={context} />
+
+    default:
+      return <InternalExternalLink link={link} handleMobileMenuOnClick={context.handleMobileMenuOnClick} />
+  }
+}
+
+interface DropdownProps {
+  item: DropDownItem
+  context: ContextProps
+}
+
+const DropDown = ({ item, context }: DropdownProps) => {
+  const { title, items } = item
+
+  return (
+    <MenuDropdown title={title}>
+      {items?.map((item, index) => {
+        const { sectionTitle, links } = item
+        return (
+          <MenuSection key={index}>
+            {sectionTitle && <MenuTitle>{sectionTitle}</MenuTitle>}
```

Comment:
Good point. index is deterministic in this case, cause we declare links in order, but I agree is better to use organic indexes.

Reply:
I've re-checked and is not worth it.

For simplicity sake, I think is better to use `index` in this case, otherwise I would need to discriminate, note that the types are unknown in this loop unless I break it into the cases. Furthermore there's cases like the "toggle dark mode" which has no url. Also, dropdown items


![image](https://user-images.githubusercontent.com/2352112/174783738-27d476dd-f1fc-426e-a2d9-81ffe4c4b2ce.png)